lib/auth/auth_with_roles.go

```diff
@@ -794,6 +794,53 @@ func (a *ServerWithRoles) PingInventory(ctx context.Context, req proto.Inventory
 	return a.authServer.PingInventory(ctx, req)
 }
 
+func (a *ServerWithRoles) GetClusterAlerts(ctx context.Context, query types.ClusterAlertQuery) ([]types.ClusterAlert, error) {
+	// unauthenticated clients can never check for alerts. we don't normally explicitly
+	// check for this kind of thing, but since alerts use an unusual access-control
+	// pattern, explicitly rejecting the nop role makes things easier.
+	if a.hasBuiltinRole(types.RoleNop) {
+		return nil, trace.AccessDenied("alerts not available to unauthenticated clients")
+	}
+
+	alerts, err := a.authServer.GetClusterAlerts(ctx, query)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// admin can see all alerts
+	if a.hasBuiltinRole(types.RoleAdmin) {
+		return alerts, nil
+	}
+
+	// filter alerts by teleport.internal labels to determine whether the alert
+	// was intended to be visible to the calling user.
+	filtered := alerts[:0]
+	for _, alert := range alerts {
+		if alert.Metadata.Labels[types.AlertPermitAll] == "yes" {
+			// alert may be shown to all authenticated users
+			filtered = append(filtered, alert)
+			continue
+		}
+
+		// TODO(fspmarshall): Support additional internal labels to help customize alert targets.
+		// maybe we could use labels to specify that an alert should only be shown to users with a
+		// specific permission (e.g. `"teleport.internal/alert-permit-permission": "node:read"`).
+		// requires further consideration.
+	}
```

Comment:
My thinking here is that whether or not an alert is appropriate to show to a given user shouldn't really be about an alert permission, but rather about whether or not the user's roles make it appropriate for the user to view alerts about the specific *subject* of the alert.

What I'm really hoping to avoid is a situation where users miss alerts that they should have seen because no one bothered to update an `alerts` RBAC option when a new alert kind was added.  E.g. if we wanted to generate alerts when a trusted cluster had been offline for more than an hour (arbitrary example), we wouldn't want to add a new RBAC option to enable receipt of that specific alert.  We'd probably just want to show that alert to anyone with `trusted_cluster:write` permissions.

Its possible that this won't be sufficient, and maybe having a permission that at least allows some users to be automatically opted into all alerts will end up being necessary.  I still think an "alert defines its target audience" model is at least theoretically better, and I'd prefer to _try_ to push that model where possible.

Reply:
Yea now it makes sense. Thanks for explanation.